### PR TITLE
Add tbptt

### PR DESCRIPTION
**Pull request by tullie**
## What does this PR do?
Adds truncated back propagation through time feature as suggested in #293

I haven't written tests or documentation yet because I wanted to make sure everyone's is happy with how it's implemented first.

### CODE REVIEW

**Comment by Borda** [comment]
@tullie pls if it is still work-in-progress add tag `[WIP]` to the name, @williamFalcon could pls add a label `wip` and app https://github.com/marketplace/wip Thx

File: pytorch_lightning/trainer/train_loop_mixin.py [hunk 1]
```diff
@@ -153,76 +154,95 @@ def run_training_batch(self, batch, batch_nb):
...
+
+        self.hiddens = None
+        for split_nb, split_batch in enumerate(splits):
+            self.split_nb = split_nb
+
+            # call training_step once per optimizer
+            for opt_idx, optimizer in enumerate(self.optimizers):
+
+                # wrap the forward step in a closure so second order methods work
+                def optimizer_closure():
```
[the remark this answers is not in the document]
Comment (williamFalcon) [reply pytorch_lightning/trainer/train_loop_mixin.py]: this function was already there. I put that there before. It's needed for lbfgs and can't be private because it's in a mixin

File: pytorch_lightning/trainer/train_loop_mixin.py [hunk 2]
```diff
@@ -153,76 +154,95 @@ def run_training_batch(self, batch, batch_nb):
...
+                    # forward pass
+                    output = self.training_forward(
+                        split_batch, batch_nb, opt_idx, self.hiddens)
+
+                    closure_loss = output[0]
```
[the remark this answers is not in the document]
Comment (williamFalcon) [reply pytorch_lightning/trainer/train_loop_mixin.py]: thanks for the feedback Borda. I actually think the way it is done already is cleaner as it's more clear what's going on.

File: pytorch_lightning/root_module/root_module.py [hunk 3]
```diff
@@ -113,6 +113,14 @@ def optimizer_step(self, epoch_nb, batch_nb, optimizer, optimizer_i, second_orde
         # clear gradients
         optimizer.zero_grad()
 
+    def tbptt_split_batch(self, batch):
```
Comment (williamFalcon) [inline pytorch_lightning/root_module/root_module.py]: @tullie is there a default implementation we can add here that would apply to 90% of use cases?

Something like:

```python
def split(batch):
   for i, x in enumerate(batch):
        # maybe recurse if we think the inputs can be lists of lists or of dicts
        if isinstance(x, list):
             x = split(x)
        
        batch[i] = some_split_fx(x)

   return batch
```

File: pytorch_lightning/trainer/train_loop_mixin.py [hunk 4]
```diff
@@ -153,76 +154,95 @@ def run_training_batch(self, batch, batch_nb):
...
+        if self.truncated_bptt:
+            if self.is_function_implemented('tbptt_batch_split'):
                 model_ref = self.get_model()
-                model_ref.backward(self.use_amp, closure_loss, optimizer)
-
-                # insert after step hook
-                if self.is_function_implemented('on_after_backward'):
+                splits = model_ref.tbptt_batch_split(batch)
+            else:
+                m = 'tpbtt_batch_split(self, batch) must be defined if truncated_bptt is True'
```
Comment (williamFalcon) [inline pytorch_lightning/trainer/train_loop_mixin.py]: love the usability thought here!

File: pytorch_lightning/root_module/root_module.py [hunk 5]
```diff
@@ -113,6 +113,14 @@ def optimizer_step(self, epoch_nb, batch_nb, optimizer, optimizer_i, second_orde
...
+        """
+        Return list of batch splits. Each split will be passed to forward_step to enable
+        truncated back propagation through time.
+        :return: list of splits
```
[the remark this answers is not in the document]
Comment (tullie) [reply pytorch_lightning/root_module/root_module.py]: Can you explain this a bit more? Where should the blank line go?

**Comment by williamFalcon** [comment]
@tullie looks great. Left feedback inline. It's already happening but just be double careful to not retain any chunks of the batch if you store those hiddens anywhere.

Did we want to do the progress bar thing we discussed for now? Add a second bar for the inner batch loop? Or should we do that in a separate PR?

File: pytorch_lightning/root_module/root_module.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Borda) [reply pytorch_lightning/root_module/root_module.py]: between 119 and 120

**Comment by Borda** [comment]
let's keep it simple and do progress bar in a separate PR

File: pytorch_lightning/root_module/root_module.py (hunk 3 above)
Comment (tullie) [reply pytorch_lightning/root_module/root_module.py]: Yeah, if we pass in the tbptt step size in the parameter then writing the default will be easier. I'll go ahead with that.

**Comment by tullie** [comment]
Yeah I thought i'd leave that for another PR, happy to look into it though. For the moment I just added the split_idx as a progress bar parameter, which was enough for my use case.

**Comment by Borda** [comment]
@tullie I believe that you can do it, but I would prefer to do in another PR. 
It is a good practice to keep PR simple and solving just one task/issue, do not add extra complexity :)

File: pytorch_lightning/root_module/root_module.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (tullie) [reply pytorch_lightning/root_module/root_module.py]: The rest of the function comments in this file don't have a newline between the descriptive text and the return. Just wanted to double check that you'd still expect this to have a new line?

[the remark this answers is not in the document]
Comment (williamFalcon) [reply pytorch_lightning/root_module/root_module.py]: not sure what line either, it looks fine to me. am i missing something? if it cleared flake8 then we're good.

[the remark this answers is not in the document]
Comment (Borda) [reply pytorch_lightning/root_module/root_module.py]: it was a minor thing, let it be and focus on the important points... :)

File: pytorch_lightning/root_module/root_module.py [hunk 6]
```diff
@@ -113,6 +113,29 @@ def optimizer_step(self, epoch_nb, batch_nb, optimizer, optimizer_i, second_orde
...
+        :return:
+        """
+        time_dims = [x.shape[1] for x in batch if isinstance(x, torch.Tensor) and x.dims() > 1]
+        assert all(x == time_dims[0] for x in time_dims), "Tensor time dimension size is ambiguous"
+
+        splits = []
+        for t in range(0, time_dims[0], split_size):
+            batch_split = []
+            for i, x in enumerate(batch):
+                if isinstance(x, torch.Tensor):
```
Comment (williamFalcon) [inline pytorch_lightning/root_module/root_module.py]: is  there a case when we still want to split non tensors? what happens,  for instance in decoding for a seq2seq ? wouldn't  we tbptt a few steps into decoding, so the targets might be python lists (ie, class labels?  word indexes, etc?)

File: docs/Trainer/hooks.md [hunk 7]
```diff
@@ -151,3 +151,27 @@ def on_after_backward(self):
...
             self.logger.experiment.add_histogram(tag=name, values=grads, global_step=self.trainer.global_step)
 ```
+
+---
+#### tbptt_split_batch
+Called in the training loop after on_batch_start if `truncated_bptt_steps > 0`. Each returned batch split is passed separately to training_step(...).
+
+```python
+def tbptt_split_batch(self, batch, split_size):
+		splits = []
```
Comment (Borda) [inline docs/Trainer/hooks.md]: I would say too many spaces...

Comment (tullie) [reply docs/Trainer/hooks.md]: Yeah good point. Just updated it.

**Comment by williamFalcon** [comment]
@tullie thanks for the big PR! great job.
